### PR TITLE
Support opfamilies/opclass for distribution in ORCA

### PR DESCRIPTION
**Pull request by hardikar**
This PR is Part 2 of the split from the previous WIP PR: #559.

Part 1 captures the Refactor commits & large MDP changes.
Part 2 captures the core commits to support opfamilies in ORCA.

All rebased now & ready to review! Please review this commit by commit - they are rather self-contained.

GPDB PR: https://github.com/greenplum-db/gpdb/pull/9401

FAQs

### How does this work?
  1. For hash joins, capture the opfamily of each hash joinable operator. Use that to create hash distribution spec requests for either side of the join.
  2. Scan operators derive a distribution spec based on opfamily (corresponding to the opclass) of each distribution column.
  3. If there is a mismatch between distribution spec requested/derived, add a Motion Redistribute node using the distribution function from the requested hash opfamily.

### What new information is captured from GPDB?
  1. Opfamily of each distribution column for every GPDB tables, external tables, CTAS tables.
    1. Needed for basic functionality.
  2. Compatible hash opfamily for each scalar operator.
    1. This finds the first hash opfamily that contains the operator.
    2. Needed for basic functionality.
  3. Default distribution (hash) opfamily of every type.
    1. Maybe NULL for some types.
    2. Needed for certain operators (e.g HashAgg) that request distribution spec that cannot be inferred in any other way: cannot derive it, cannot get it from any scalar op etc.
  4. Legacy opfamilies for types & scalar operators.
    1. Needed for supporting tables distributed by legacy opclasses.

### Did you add any traceflags? What are they for?
  1. EopttraceConsiderOpfamiliesForDistribution:
    1. GPDB6 enables this by default.
    2. If this is off, opfamilies is ignored and set to NULL.
    3. This should mimic behavior before this PR. Ctest MDPs are run this way.
  2. EopttraceUseLegacyOpfamilies:
    1. Set if ANY distribution col in the query uses a legacy opfamily/opclass.
    2. The MDCache captures both legacy and new opfamilies.
    3. Code returns legacy opfamilies instead of the default when this is set
### Assumptions made in the PR:
  1. In the distribution spec,
    1. Opfamilies arrays must be populated (if EopttraceConsiderOpfamiliesForDistribution is set) or NULL (if unset)
    2. Opfamilies array must not contain a NULL element.
  2. In PdsCopyWithRemappedColumns(), remapping columns will not change opfamily
  3. Gp\_use\_legacy\_hashops GUC is irrelevant in this effort.5
  4. When determining equiv classes, only consider operators that apply on the default opfamilies of the left type. (Needed for some CDistributionSpecHashed matching functions)

### What work is left?
  [ ] Add any missing assertions, throw exceptions & add tests! So highlight this if you see it!
  [ ] Refactor to use a factory method for CDistributionSpecHashed.

### CODE REVIEW

File: libnaucrates/src/parser/CParseHandlerMDRelation.cpp [hunk 1]
```diff
@@ -251,12 +263,19 @@ CParseHandlerMDRelation::EndElement
...
+
 	md_col_array->AddRef();
 	md_index_info_array->AddRef();
  	mdid_triggers_array->AddRef();
  	mdid_check_constraint_array->AddRef();
 
+	IMdIdArray *distr_opfamilies = NULL;
+	if (m_rel_distr_policy == IMDRelation::EreldistrHash && m_opfamilies_parse_handler != NULL)
+	{
+		distr_opfamilies = dynamic_cast<CParseHandlerMetadataIdList*>(m_opfamilies_parse_handler)->GetMdIdArray();
```
Comment (zellerh) [inline libnaucrates/src/parser/CParseHandlerMDRelation.cpp]: Use a static cast instead to save a few femtoseconds?

File: libgpopt/src/operators/CPhysicalHashJoin.cpp [hunk 2]
```diff
@@ -150,17 +160,34 @@ CPhysicalHashJoin::CreateHashRedistributeRequests
...
+			CDistributionSpecHashed *pdshashedCurrent = GPOS_NEW(mp) CDistributionSpecHashed(pdrgpexprCurrent,
+																							 true /* fNullsCollocated */,
+																							 opfamilies);
 			m_pdrgpdsRedistributeRequests->Append(pdshashedCurrent);
 		}
 	}
 	// add a request that contains all hash join keys
 	pdrgpexpr->AddRef();
-	CDistributionSpecHashed *pdshashed = GPOS_NEW(mp) CDistributionSpecHashed(pdrgpexpr, true /* fNullsCollocated */);
+	if (GPOS_FTRACE(EopttraceConsiderOpfamiliesForDistribution)) // FIGGY
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/src/operators/CPhysicalHashJoin.cpp]: Oops I missed this one :)

File: libgpopt/src/xforms/CXformUtils.cpp [hunk 3]
```diff
@@ -2551,7 +2551,8 @@ CXformUtils::LookupJoinKeys
 	CMemoryPool *mp,
 	CExpression *pexpr,
 	CExpressionArray **ppdrgpexprOuter,
-	CExpressionArray **ppdrgpexprInner
+	CExpressionArray **ppdrgpexprInner,
+	IMdIdArray **join_opfamilies
 	)
 {
 	GPOS_ASSERT(NULL != ppdrgpexprOuter);
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/src/xforms/CXformUtils.cpp]: Good idea! done!

File: libnaucrates/include/naucrates/traceflags/traceflags.h [hunk 4]
```diff
@@ -187,6 +187,13 @@ namespace gpos
 
 		// Expand LOJs in N-aryjoin
 		EopttraceEnableLOJInNAryJoin = 103033,
+
+		// Support Opfamilies in distribution specs
+		EopttraceConsiderOpfamiliesForDistribution = 103034,
+
+		// Support Opfamilies in distribution specs
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libnaucrates/include/naucrates/traceflags/traceflags.h]: Done

File: libgpopt/src/operators/CPhysicalUnionAll.cpp [hunk 5]
```diff
@@ -104,14 +104,56 @@ CPhysicalUnionAll::CPhysicalUnionAll
...
+		{
+			CColRef *colref = (*colref_array)[ulCol];
+			CExpression *pexpr = CUtils::PexprScalarIdent(mp, colref);
+			pdrgpexpr->Append(pexpr);
+		}
+
+		// create a hashed distribution on input columns of the current child
+		BOOL fNullsColocated = true;
+		CDistributionSpec *pdshashed =
+			CDistributionSpecHashed::MakeHashedDistrSpec(mp, pdrgpexpr, fNullsColocated, NULL, NULL);
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/src/operators/CPhysicalUnionAll.cpp]: I used `opfamilies=NULL` as scaffold. I intended to get rid of it later, but forgot. Thanks for pointing it out.

File: libnaucrates/src/operators/CDXLOperatorFactory.cpp [hunk 6]
```diff
@@ -1480,8 +1480,10 @@ CDXLOperatorFactory::MakeDXLHashExpr
 							(
 							dxl_memory_manager,
 							attrs,
-							EdxltokenTypeId,
-							EdxltokenScalarHashExpr
+							EdxltokenOpfamily,
+							EdxltokenScalarHashExpr,
+							true /* is_optional */,
+							NULL
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libnaucrates/src/operators/CDXLOperatorFactory.cpp]: The default value.

File: libgpopt/src/operators/CPhysical.cpp [hunk 7]
```diff
@@ -310,7 +310,20 @@ CPhysical::PdsCompute
...
 
-			pds = GPOS_NEW(mp) CDistributionSpecHashed(pdrgpexpr, true /*fNullsColocated*/);
+			IMdIdArray *opfamilies = NULL;
+			if (GPOS_FTRACE(EopttraceConsiderOpfamiliesForDistribution))
+			{
+				opfamilies = GPOS_NEW(mp) IMdIdArray(mp);
+				for (ULONG ul = 0; ul < size; ul++)
+				{
+					IMDId *opfamily = (*ptabdesc->DistrOpfamilies())[ul];
+					opfamily->AddRef();
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/src/operators/CPhysical.cpp]: Sure.

File: libgpopt/src/base/CConstraint.cpp [hunk 8]
```diff
@@ -425,6 +428,21 @@ CConstraint::PcnstrFromScalarCmp
 			return NULL;
 		}
 
+		if (GPOS_FTRACE(EopttraceConsiderOpfamiliesForDistribution))
+		{
+			CMDAccessor *mda = COptCtxt::PoctxtFromTLS()->Pmda();
+			CScalarCmp *sc_cmp = CScalarCmp::PopConvert(pexpr->Pop());
+			const IMDScalarOp *op = mda->RetrieveScOp(sc_cmp->MdIdOp());
+
+			IMDId *left_mdid = CScalar::PopConvert(pexprLeft->Pop())->MdidType();
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/src/base/CConstraint.cpp]: Honestly, it's probably better to check both. I'll update the code.

File: data/dxl/minidump/JoinCitextVarchar.mdp [hunk 9]
```diff
@@ -26,7 +26,7 @@
         </dxl:CostParams>
       </dxl:CostModelConfig>
       <dxl:Hint MinNumOfPartsToRequireSortOnInsert="2147483647" JoinArityForAssociativityCommutativity="18" ArrayExpansionThreshold="100" JoinOrderDynamicProgThreshold="10" BroadcastThreshold="100000" EnforceConstraintsOnDML="false" PushGroupByBelowSetopThreshold="10"/>
-      <dxl:TraceFlags Value="102074,102120,102146,103001,103014,103022,103027,103029,103037,104002,104003,104004,104005,105000,106000"/>
+      <dxl:TraceFlags Value="102074,102120,102146,103001,103014,103022,103027,103029,103038,104002,104003,104004,104005,105000,106000"/>
```
Comment (ashuka24) [inline data/dxl/minidump/JoinCitextVarchar.mdp]: why was this a traceflag change and not a traceflag addition? Do we no longer want `EopttraceAllowGeneralPredicatesforDPE`? Same question for these mdps as well: 
```
data/dxl/minidump/JoinDefaultOpfamiliesUsingNonDefaultOpfamilyOp.mdp
data/dxl/minidump/JoinTinterval.mdp
data/dxl/minidump/MultiColumnAggWithDefaultOpfamilies.mdp
```

File: libgpopt/include/gpopt/xforms/CXformUtils.h [hunk 10]
```diff
@@ -1403,9 +1403,9 @@ namespace gpopt
 				if (GPOS_FTRACE(EopttraceConsiderOpfamiliesForDistribution))
 				{
 					CMDAccessor *mda = COptCtxt::PoctxtFromTLS()->Pmda();
-					const IMDScalarOp *scop = mda->RetrieveScOp(mdid_scop);
-					IMDId *hash_opfamily = scop->HashOpfamiliyMdid();
+					IMDId *hash_opfamily = mda->RetrieveScOp(mdid_scop)->HashOpfamiliyMdid();
```
Comment (ashuka24) [inline libgpopt/include/gpopt/xforms/CXformUtils.h]:
```suggestion
					IMDId *hash_opfamily = mda->RetrieveScOp(mdid_scop)->HashOpfamilyMdid();
```

File: data/dxl/minidump/JoinCitextVarchar.mdp (hunk 9 above)
Comment (hardikar) [reply data/dxl/minidump/JoinCitextVarchar.mdp]: It's cuz I create these MDP using 103037. Which needed to be changed to 103038 after rebase. All of these are new MDPs.

File: libgpopt/include/gpopt/xforms/CXformUtils.h (hunk 10 above)
Comment (hardikar) [reply libgpopt/include/gpopt/xforms/CXformUtils.h]: Fixed.

**Comment by hardikar** [comment]
Closed in favor of https://github.com/greenplum-db/gpdb/pull/9889.

**Pull request closed without merging**